src/java/com/palantir/cassandra/utils/MutationVerificationUtils.java

```
@@ -0,0 +1,63 @@
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.cassandra.db.Keyspace;
+import org.apache.cassandra.db.Mutation;
+import org.apache.cassandra.dht.Token;
+import org.apache.cassandra.locator.NetworkTopologyStrategy;
+import org.apache.cassandra.service.StorageService;
+import org.apache.cassandra.utils.FBUtilities;
+
+public class MutationVerificationUtils
+{
+    private static final boolean VERIFY_KEYS_ON_WRITE = Boolean.valueOf(System.getProperty("palantir_cassandra.verify_keys_on_write", "false"));
+    private static final Logger logger = LoggerFactory.getLogger(MutationVerificationUtils.class);
+
+    private MutationVerificationUtils() {}
+
+    public static void verifyMutation(Mutation mutation) {
+        if (VERIFY_KEYS_ON_WRITE) {
+            Keyspace keyspace = Keyspace.open(mutation.getKeyspaceName());
+            if (keyspace.getReplicationStrategy() instanceof NetworkTopologyStrategy) {
+                Token tk = StorageService.getPartitioner().getToken(mutation.key());
+                List<InetAddress> naturalEndpoints = StorageService.instance.getNaturalEndpoints(mutation.getKeyspaceName(), tk);
+                Collection<InetAddress> pendingEndpoints = StorageService.instance.getTokenMetadata().pendingEndpointsFor(tk, mutation.getKeyspaceName());
+
+                if(!naturalEndpoints.contains(FBUtilities.getBroadcastAddress()) && !pendingEndpoints.contains(FBUtilities.getBroadcastAddress())) {
+                    keyspace.metric.invalidMutations.inc();
+                    logger.error("Cannot apply mutation as this host {} does not contain key {}. Only hosts {} and {} do.",
+                                 FBUtilities.getBroadcastAddress(),
+                                 mutation.key(),
+                                 naturalEndpoints,
+                                 pendingEndpoints);
+                    throw new RuntimeException("Cannot apply mutation as this host does not contain key.");
```

i assume the outcome here is that we fail the write?

---

Yes! And thinking through QUORUM and ALL, QUORUM requests should still succeed, as we'd expect at worst only one node to be viewed incorrectly here, and for ALL, this would only be broken "temporarily" until gossip caught up.

---

In the case that it is more than one node, the current state of the coordinator is so bad that we actually do want to fail hard